### ISSUE TITLE
Simple question about using the hard boundary conditions

### ISSUE THREAD

**Issue by akaptano** [lululxvi/deepxde#1011]
Hello,

I am trying to understand how the hard boundary is implemented in the DeepXDE code. There are some nice examples, but I am struggling how to explicitly check that the hard BCs are satisfied during training. For instance, I tried using loss_weights=[1, 1] so that I can see the output of the PDE and BC losses, but the BC loss looks nonzero even after the apply_output_transform. So I'm guessing that either (1) I am misunderstanding what loss terms get printed during training or (2) It is the boundary loss term of y that gets printed instead of the boundary loss of output_transform(y), so it is nonzero. If so, how do you recommend a careful check of the hard BCs for more sophisticated custom boundary conditions? 

Full code is a simple version of the Poisson 1D example. 

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import matplotlib.pyplot as plt
import numpy as np

import deepxde as dde
from deepxde.backend import tf

def main():
    def pde(x, y):
        dy_xx = dde.grad.hessian(y, x)
        return -dy_xx - np.pi ** 2 * tf.sin(np.pi * x)

    def boundary(x, on_boundary):
        return on_boundary

    def func(x):
        return np.sin(np.pi * x)

    geom = dde.geometry.Interval(-1, 1)
    bc = dde.DirichletBC(geom, func, boundary)
    data = dde.data.PDE(geom, pde, bc, 16, 2, solution=func, num_test=100)

    layer_size = [1] + [50] * 3 + [1]
    activation = "tanh"
    initializer = "Glorot uniform"
    net = dde.maps.FNN(layer_size, activation, initializer)

    def output_transform(x, y):
        return x * (np.pi - x) * y + x

    net.apply_output_transform(output_transform)

    model = dde.Model(data, net)
    model.compile("adam", lr=0.001, metrics=["l2 relative error"], loss_weights=[1, 1])

    checkpointer = dde.callbacks.ModelCheckpoint(
        "model/model.ckpt", verbose=1, save_better_only=True
    )
    # ImageMagick (https://imagemagick.org/) is required to generate the movie.
    movie = dde.callbacks.MovieDumper(
        "model/movie", [-1], [1], period=100, save_spectrum=True, y_reference=func
    )
    losshistory, train_state = model.train(
        epochs=10000, callbacks=[checkpointer, movie]
    )

    dde.saveplot(losshistory, train_state, issave=True, isplot=True)

    # Plot PDE residual
    model.restore("model/model.ckpt-" + str(train_state.best_step), verbose=1)
    x = geom.uniform_points(1000, True)
    y = model.predict(x, operator=pde)
    plt.figure()
    plt.plot(x, y)
    plt.xlabel("x")
    plt.ylabel("PDE residual")
    plt.show()

if __name__ == "__main__":
    main()

**Comment by akaptano** [comment]
By the way, it prints two loss terms even when I set bc = [] so I am also confused in this situation what the second loss term in the training is.

**Comment by forxltk** [comment]
1. The BC should be $u(-1)=u(1)=0$. But your `output_transform` is $u(0)=0, u(\pi)=\pi$. It should be `(1 + x) * (1 - x) * y`
2. Because you set `loss_weights=[1, 1]`.
